politeiad/backendv2/tstorebe/tstore/tlogclient.go

```diff
@@ -0,0 +1,789 @@
+func (t *tclient) leavesByRange(treeID int64, startIndex, count int64) ([]*trillian.LogLeaf, error) {
+	log.Tracef("trillian leavesByRange: %v %v %v", treeID, startIndex, count)
+
+	glbrr, err := t.log.GetLeavesByRange(t.ctx,
+		&trillian.GetLeavesByRangeRequest{
+			LogId:      treeID,
+			StartIndex: startIndex,
+			Count:      count,
+		})
+	if err != nil {
+		return nil, err
+	}
+
+	return glbrr.Leaves, nil
+}
+
+// LeavesAll returns all of the leaves for the provided treeID.
+//
+// This function satisfies the tlogClient interface.
+func (t *tclient) LeavesAll(treeID int64) ([]*trillian.LogLeaf, error) {
+	log.Tracef("trillian LeavesAll: %v", treeID)
+
+	// Get tree
+	tree, err := t.Tree(treeID)
+	if err != nil {
+		return nil, fmt.Errorf("tree: %v", err)
+	}
+
+	// Get signed log root
+	_, lr, err := t.SignedLogRoot(tree)
+	if err != nil {
+		return nil, fmt.Errorf("SignedLogRoot: %v", err)
+	}
+	if lr.TreeSize == 0 {
+		return []*trillian.LogLeaf{}, nil
+	}
+
+	// Get all leaves
+	leaves, err := t.leavesByRange(treeID, 0, int64(lr.TreeSize))
+	if err != nil {
+		return nil, fmt.Errorf("leavesByRange: %v", err)
+	}
+
+	return leaves, nil
+}
+
+// Close closes the trillian grpc connection.
+//
+// This function satisfies the tlogClient interface.
+func (t *tclient) Close() {
+	log.Tracef("trillian Close %v", t.host)
+
+	t.grpc.Close()
+}
+
+// merkleLeafHash returns the merkle leaf hash for the provided leaf value.
+// This is the same merkle leaf hash that is calculated by trillian.
+func merkleLeafHash(leafValue []byte) []byte {
+	return hasher.HashLeaf(leafValue)
+}
+
+func newLogLeaf(leafValue []byte, extraData []byte) *trillian.LogLeaf {
+	return &trillian.LogLeaf{
+		LeafValue: leafValue,
+		ExtraData: extraData,
+	}
+}
+
+func newTrillianKey() (crypto.Signer, error) {
+	return keys.NewFromSpec(&keyspb.Specification{
+		Params: &keyspb.Specification_Ed25519Params{},
+	})
+}
+
+// newTClient returns a new tclient.
+func newTClient(host, keyFile string) (*tclient, error) {
+	// Setup trillian key file
+	if !util.FileExists(keyFile) {
+		// Trillian key file does not exist. Create one.
+		log.Infof("Generating trillian private key")
+		key, err := newTrillianKey()
+		if err != nil {
+			return nil, err
+		}
+		b, err := der.MarshalPrivateKey(key)
+		if err != nil {
+			return nil, err
+		}
+		err = ioutil.WriteFile(keyFile, b, 0400)
+		if err != nil {
+			return nil, err
+		}
+		log.Infof("Trillian private key created: %v", keyFile)
+	}
+
+	// Default gprc max message size is ~4MB (4194304 bytes). This is
+	// not large enough for trees with tens of thousands of leaves.
+	// Increase it to 20MB.
+	maxMsgSize := grpc.WithMaxMsgSize(20 * 1024 * 1024)
+
+	// Setup trillian connection
+	g, err := grpc.Dial(host, grpc.WithInsecure(), maxMsgSize)
+	if err != nil {
+		return nil, fmt.Errorf("grpc dial: %v", err)
+	}
+
+	// Load trillian key pair
+	var privateKey = &keyspb.PrivateKey{}
+	privateKey.Der, err = ioutil.ReadFile(keyFile)
```

Comment:
This has to come out of an environment variable or kv store and passed in instead of using a file.

Reply:
Done